s3-store/src/main.rs

```diff
@@ -0,0 +1,349 @@
+    BlobFormat,
+};
+use iroh_io::{AsyncSliceReaderExt, HttpAdapter};
+use iroh_net::{key::SecretKey, ticket::BlobTicket, MagicEndpoint, NodeAddr};
+use iroh_s3_bao_store::S3Store;
+use serde::Deserialize;
+use std::{
+    fmt::{Display, Formatter},
+    str::FromStr,
+    sync::Arc,
+    time::Duration,
+};
+use tokio_util::task::LocalPoolHandle;
+use url::Url;
+
+/// Send a file or directory between two machines, using blake3 verified streaming.
+///
+/// For all subcommands, you can specify a secret key using the IROH_SECRET
+/// environment variable. If you don't, a random one will be generated.
+///
+/// You can also specify a port for the magicsocket. If you don't, a random one
+/// will be chosen.
+#[derive(Parser, Debug)]
+pub struct Args {
+    #[clap(subcommand)]
+    pub command: Commands,
+}
+
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+pub enum Format {
+    #[default]
+    Hex,
+    Cid,
+}
+
+impl FromStr for Format {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_ascii_lowercase().as_str() {
+            "hex" => Ok(Format::Hex),
+            "cid" => Ok(Format::Cid),
+            _ => Err(anyhow::anyhow!("invalid format")),
+        }
+    }
+}
+
+impl Display for Format {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Format::Hex => write!(f, "hex"),
+            Format::Cid => write!(f, "cid"),
+        }
+    }
+}
+
+fn print_hash(hash: &blake3::Hash, format: Format) -> String {
+    match format {
+        Format::Hex => hash.to_hex().to_string(),
+        Format::Cid => hash.to_string(),
+    }
+}
+
+#[derive(Subcommand, Debug)]
+pub enum Commands {
+    /// Send a file or directory.
+    ServeS3(ServeS3Args),
+
+    /// Receive a file or directory.
+    ServeUrls(ImportS3Args),
+}
+
+#[derive(Parser, Debug)]
+pub struct CommonArgs {
+    /// The port for the magic socket to listen on.
+    ///
+    /// Defauls to a random free port, but it can be useful to specify a fixed
+    /// port, e.g. to configure a firewall rule.
+    #[clap(long, default_value_t = 0)]
+    pub magic_port: u16,
+
+    #[clap(long, default_value_t = Format::Hex)]
+    pub format: Format,
+
+    #[clap(short = 'v', long, action = clap::ArgAction::Count)]
+    pub verbose: u8,
+}
+
+#[derive(Parser, Debug)]
+pub struct ServeS3Args {
+    /// Url to the s3 bucket root.
+    pub url: Url,
+
+    /// Top level directory name.
+    #[clap(long)]
+    pub name: Option<String>,
+
+    #[clap(flatten)]
+    pub common: CommonArgs,
+}
+
+#[derive(Parser, Debug)]
+pub struct ImportS3Args {
+    /// Url to the s3 bucket root.
+    pub url: Vec<Url>,
+
+    #[clap(flatten)]
+    pub common: CommonArgs,
+}
+
+/// Get the secret key or generate a new one.
+///
+/// Print the secret key to stderr if it was generated, so the user can save it.
+fn get_or_create_secret(print: bool) -> anyhow::Result<SecretKey> {
```

Comment:
we need to add this functionality to the `Node::builder`, this code has been duplicated at least 10 times by now..

Reply:
not the printing, but the loading, same with the loading from a file, maybe sth like

- `SecretKey::from_env(env_variable) -> Result<Option<Self>>`
- `SecretKey::from_file(path) -> Result<Option<Self>>`

and then it can be used like this

```rust
let secret_key = SecretKey::from_env("IROH_SECRET_KEY")?.unwrap_or_else(SecretKey::generate);
```